jicofo-common/src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoomImpl.kt

```diff
@@ -0,0 +1,769 @@
+    }
+
+    /** The number of members that currently have their audio sources unmuted. */
+    private var numAudioSenders = 0
+
+    /** The number of members that currently have their video sources unmuted. */
+    private var numVideoSenders = 0
+
+    override fun getRoomJid() = roomJid
+    override fun getXmppProvider() = xmppProvider
+    override fun isBreakoutRoom() = isBreakoutRoom
+    override fun getMainRoom() = mainRoom
+    override fun getMeetingId(): String? = meetingId
+    override fun addListener(listener: ChatRoomListener) = eventEmitter.addHandler(listener)
+    override fun removeListener(listener: ChatRoomListener) = eventEmitter.removeHandler(listener)
+    override fun isJoined() = muc.isJoined
+    override fun getMembers(): List<ChatRoomMember> = synchronized(members) { return members.values.toList() }
+    override fun getChatMember(occupantJid: EntityFullJid) = members[occupantJid]
+    override fun getMembersCount() = members.size
+    override fun isAvModerationEnabled(mediaType: MediaType) = avModeration.isEnabled(mediaType)
+    override fun setAvModerationEnabled(mediaType: MediaType, value: Boolean) =
+        avModeration.setEnabled(mediaType, value)
+    override fun setAvModerationWhitelist(mediaType: MediaType, whitelist: List<String>) =
+        avModeration.setWhitelist(mediaType, whitelist)
+    override fun isMemberAllowedToUnmute(jid: Jid, mediaType: MediaType): Boolean =
+        avModeration.isAllowedToUnmute(mediaType, jid.toString())
+
+    // Use toList to avoid concurrent modification. TODO: add a removeAll to EventEmitter.
+    override fun removeAllListeners() = eventEmitter.eventHandlers.toList().forEach { eventEmitter.removeHandler(it) }
+
+    fun setStartMuted(startMuted: BooleanArray) = eventEmitter.fireEvent {
+        startMutedChanged(startMuted[0], startMuted[1])
+    }
+
+    @Throws(SmackException::class, XMPPException::class, InterruptedException::class)
+    override fun join() {
+        // TODO: clean-up the way we figure out what nickname to use.
+        resetState()
+        joinAs(xmppProvider.config.username)
+    }
+
+    /**
+     * Prepare this [ChatRoomImpl] for a call to [.joinAs], which send initial presence to
+     * the MUC. Resets any state that might have been set the previous time the MUC was joined.
+     */
+    private fun resetState() {
+        synchronized(members) {
+            if (members.isNotEmpty()) {
+                logger.warn("Removing " + members.size + " stale members.")
+                members.clear()
+            }
+        }
+        synchronized(this) {
+            role = null
+            lastPresenceSent = null
+            meetingId = null
+            logger.addContext("meeting_id", "")
+            isBreakoutRoom = false
+            mainRoom = null
+            avModeration.reset()
+        }
+    }
+
+    @Throws(SmackException::class, XMPPException::class, InterruptedException::class)
```

Comment:
Do you still need the `@Throws` declaration in Kotlin?

Reply:
I don't think it's strictly needed, but it's nice to have IMO